### DISCUSSION TITLE
Hail Mary mode

### DISCUSSION THREAD

**Discussion by Tenfilip** [MoonRide303/Fooocus-MRE#133] [Ideas]
Hi,

Thanks for great fork! 

My idea is: is it possible to introduce "Hail Mary" (inspired by Armitage from Kali linux) mode for styles? 

1/ I setup the tool as I always do, but I check the Hail Mary box.
2/ Click generate.
3/ Tool generates as normal, but for each of the styles. 

This would help to understand the styles, but also find the right one for the job - **would also be fun!**

**Comment by DrJKL** [comment]
How do you feel about "Tasting Menu" or "Sampling Platter"?

**Reply by Tenfilip** [reply]
I'm pretty agnostic regarding the name, I care about the functionality. I thought it would be cool to use this name that already is used in another cool OSS product, but any other would do, really. Maybe a simple "Run for each style"?

**Reply by DrJKL** [reply]
What's the other product?

**Reply by Tenfilip** [reply]
I meant the Armitage, in Kali Linux.

**Comment by barepixels** [comment]
@runew0lf over at [RuinedFooocus](https://github.com/runew0lf/RuinedFooocus) have came up with a solution.  Using the **multiple prompt feature** and **add style:stylename to prompt to set the required style** feature in [RuinedFooocus](https://github.com/runew0lf/RuinedFooocus) one can batch render an image for each style.  While not as elegance or easy as one-button it does work

Read more about this here:  https://github.com/runew0lf/RuinedFooocus/issues/20

**Comment by ptits** [comment]
Hello all,
i am not good with repos and forks, so I just modify couple of lines in async_worker.py

**logic is following
if image_number (taken from UI) is 128 Generate Button renders just selected styles, one by one, no mixing, 8 max (you can tweak 8 in code)
if image_number (taken from UI) is 127 Generate Button renders ALL THE styles, one by one, no mixing (208 for my style set)**

I dont want to hassle with additional check boxes in UI (they will disapear with next update)

You can download py file it from here and put into Fooocus-MRE/modules
https://drive.google.com/file/d/10d7RkHAwAs65aasj1mpd1PWcQ7X0nkof/view?usp=sharing

It will work until first update (so you might want to avoid update to play with it).

You can also modify it yourself:

find the line:
        for i in range(image_number):         
and replace it with lines:

        renderStyles = 0
        from modules.sdxl_styles import style_keys
        style_sel = style_selections
        if image_number == 128:
            image_number = len(style_sel)
            renderStyles = 1
        if image_number == 127:
            style_sel = style_keys
            image_number = len(style_sel)
            renderStyles = 2
        for i in range(image_number):
            if renderStyles: style_selections = [style_sel[i]]



need to make cheats for output log but too lazy,

**Comment by barepixels** [comment]
please ignore

**Reply by ptits** [reply]
No, no.

Redo it, but skip step 3. Just replace async_worker.ry

Then run Fooocus

Set Image Number to 127 in the UI (Settings tab).

Push Generate.

127 will tell to generate ALL styles.

No need to modify code. No need to select Styles.

127 is secret trigger to run cycle.

**Reply by barepixels** [reply]
Ok doing it now.  BRB

**Comment by barepixels** [comment]
**Confirm working**  Thank you @ptits
This is what I did.
1) rename orginal async_worker.py to ORGasync_worker.py as backup
2) place your async_worker.py in modules folder
4) make sure only "**sdxl_styles_mre.json**" in sdxl_styles folder (Keeping it simple for testing and doing small batch at a time)
5) launch Fooocus-MRE using "**run-mre-no-update.bat**"
6) From the settings tab: Quality On, Batch **Image number = 127**, Prompt expansion OFF, Seed = 0, not random 
7) From Models tab, default models, refiner, loras
8) press Generate
9) develop a smile, and start dancing
10) open **sdxl_styles_mre.json**, copying each style name and renaming the images as it is generated

**Reply by ptits** [reply]
Next dance is:
Select 8 styles you want
Set Image Number to 128.
Push Generate.
It will render 8 selected styles.

128 is secret number to tell focus to render only selected styles.

No more secrets))

It will work with any other settings

**Comment by barepixels** [comment]
Confirm Set Image Number to 128 works great.  @ptits you made my day.  Now just need @MoonRide303 to merge this and or improve upon **Run All Styles**

**Reply by ptits** [reply]
I can also make automatic renaming for output files, making them "style+random number" if you need

**Reply by barepixels** [reply]
That would save a lot of time.  Manually renaming each file by cross-referencing a json is a pain in the ass.  Can you do **model-style-seed** for the file name?  If not it's OK, I can use a batch rename tool and add the model to the front of the file name

**Comment by ptits** [comment]
Here it is. Try it
https://drive.google.com/file/d/1zux9znjVK2Md_n0wh1RMhev0jLzrs2LB/view?usp=sharing

**Reply by barepixels** [reply]
OSError

![error1](https://github.com/MoonRide303/Fooocus-MRE/assets/92839033/40ed4e55-2058-49f7-ad37-33710254fd54)


OSError: [Errno 22] Invalid argument: 'E:\\Fooocus-MRE\\Fooocus-MRE\\outputs\\2023-09-30\\sd_xl_base_1.0_0.9vae|Default (Slightly Cinematic)|0|23.6990".jpg'

Windows doesn't like  | (pipe) and " (double quote) in the file name
**sd_xl_base_1.0_0.9vae|Default (Slightly Cinematic)|0|23.6990".jpg**

UPDATE:  same error with batch size set at 1 not 128

**Reply by barepixels** [reply]
UPDATE: I managed to fix it.

async_worker.py line 603
Before:
`fileStyleName = f'{base_model_name.replace(".safetensors", "")}|{style_sel[current_task_idx]}|{task["task_seed"]}|{execution_time:.4f}".ext'

After:
`fileStyleName = f'{base_model_name.replace(".safetensors", "")}-{style_sel[current_task_idx]}-{task["task_seed"]}-{execution_time:.4f}.ext'`

Change | (pipe) to - (dash)  and remove one " (double quote) by .jpg

![list](https://github.com/MoonRide303/Fooocus-MRE/assets/92839033/227f449b-4151-417f-a8a3-4791ac3623c1)

SIDE EFFECT: this will be the new file name convention for all generated images, I kinda like it

**Reply by ptits** [reply]
Nice to see you make it work!
I am on Ubuntu, so next time I will check windows execution also.

**Reply by ptits** [reply]
@barepixels 
As long as new update can nicely iterate through all styles, I made quick solution for naming images as you need.
I found private_logger.py and insert some code, you will see it at the beginning, feel free to modify
private_logger.py is not core file - it is for logging and file renaming, easу for tweaks, just replace original one in \modules
Let me know if it is work for you
https://drive.google.com/file/d/1LR2inGy2pdPI39ef6a9njuMx1MU7MSC1/view?usp=sharing

**Comment by MoonRide303** [comment]
@Tenfilip Added in v2.0.78.5 as Style Iterator. Can be used to check combinations with other styles, too (if any is selected). Information about style is included in metadata (if you enable saving it), so it shouldn't be necessary to include style name in file name.

**Reply by ptits** [reply]
@MoonRide303 Thank you so much! Works perfect.
Combination trick is very elegant.

About renaming files:
When doing tests we usually browse images somewhere and using filenames is saving a lot of time, for example find all images with particular style or sort images with particular model. I also use simple py script to put filename on the bottom of image itself.

Yes, it is possible to make script for renaming images using metadata but having an option to keep images name in form of {model}-{style}-{seed}-{rand} will be huge time saver (for eyes and brains).

May be using Keep Input Names can trigger {model}-{style}-{seed}-{rand} naming?

**Comment by barepixels** [comment]
It would be a tedious task to manually get the style from metadata and rename files.  Imagine having to do 200 styles for each model.  Let's say you have 10 models, that would be 2000 files to manually rename for the prompt "a cat"  then manually rename 2000 files for "landscape", and then manually rename 2000 files for "a naked lady"  **HELP, HELP, HELP**

The way I see it there are 4 possible solutions

1) an external script that scans the metadata and renames the files within a folder
2) in settings.json, the option for file name convention  a) default) b) **extended** {model}-{style}-{seed}-{rand}
3) auto switch to **extended file name** `{style}__{model}__{seed}` when use **Style Iterator**
4) have an extra async_worker.py.  One with the extended file name.  User can swap the async_worker.py and launch with **run-mre-no-update.bat** when plan to use **Style Iterator**

My vote would be number 3

**Reply by ptits** [reply]
@barepixels 
As long as new update can nicely iterate through all styles, I made quick solution for naming images as you need.
I found private_logger.py and insert some code, you will see it at the beginning, feel free to modify
private_logger.py is not core file - it is for logging and file renaming, easу for tweaks, just replace original one in \modules
Let me know if it is work for you
https://drive.google.com/file/d/1LR2inGy2pdPI39ef6a9njuMx1MU7MSC1/view?usp=sharing

**Reply by barepixels** [reply]
@ptits it works great.  One thing people should know, when you are **not** using **Style Iterator** the file name still stays in the extended format (Default (Slightly Cinematic)__Seed-0__sd_xl_base_1.0_0.9vae)   

You can keep this as **always** OR swap back to original **private_logger.py** when done using **Style Iterator**

Wishing @MoonRide303 will make Fooocus **auto switch** to extended file name `{style}__{model}__{seed}` when using Style Iterator

**Reply by ptits** [reply]
@barepixels 
Here is private_logger.py with the following rename notation: Style_Time_Model
Default (Slightly Cinematic)__Time-14-14-15__sd_xl_base_1.0_0.9vae.jpg
You can easily read and correct code for you needs. Like adding Date.

https://drive.google.com/file/d/11w4zsIMwOAw_eyfE4YYW0_ZkNtqHpjP_/view?usp=sharing

**Reply by barepixels** [reply]
thank you brother

**Comment by Tenfilip** [comment]
It seems to work as intended, I also like the name Style Iterator. Thank you!
Update: I ran a few tests and I'm really happy with it, it's an easy way to fuzz a nice image.

**Comment by 3Diva** [comment]
@MoonRide303 The "Style Iterator" is SO COOL! Thank you to everyone who worked on this! It's a HUGE time saver! You guys are awesome!
